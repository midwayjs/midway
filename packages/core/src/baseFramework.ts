import {
  IConfigurationOptions,
  IMidwayApplication,
  IMidwayBootstrapOptions,
  IMidwayContainer,
  IMidwayFramework,
  MidwayFrameworkType,
  MidwayProcessTypeEnum,
} from './interface';
import { ContainerLoader } from './';
import { APPLICATION_KEY } from '@midwayjs/decorator';

export abstract class BaseFramework<
  APP extends IMidwayApplication,
  T extends IConfigurationOptions
> implements IMidwayFramework<APP, T> {
  protected isTsMode = true;
  protected baseDir: string;
  protected appDir: string;
  protected containerLoader: ContainerLoader;
  public configurationOptions: T;
  public app: APP;

  public configure(options: T): BaseFramework<APP, T> {
    this.configurationOptions = options;
    return this;
  }

  public async initialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {
    this.baseDir = options.baseDir;
    this.appDir = options.appDir;

    this.containerLoader = new ContainerLoader({
      baseDir: this.baseDir,
      isTsMode: this.isTsMode,
      preloadModules: options.preloadModules || [],
    });

    /**
     * initialize containerLoader and initialize ioc container instance
     */
    await this.beforeInitialize(options);
    this.containerLoader.initialize();

    /**
     * load directory and bind files to ioc container
     */
    await this.beforeDirectoryLoad(options);
    const applicationContext = this.containerLoader.getApplicationContext();
    applicationContext.registerObject('baseDir', this.baseDir);
    applicationContext.registerObject('appDir', this.appDir);
    // 如果没有关闭autoLoad 则进行load
    this.containerLoader.loadDirectory(options);

    // register app
    this.containerLoader.registerHook(APPLICATION_KEY, () => {
      return this.getApplication();
    });

    await this.afterDirectoryLoad(options);

    if (!this.app.getApplicationContext) {
      this.app = this.defineApplicationProperties(this.app);
    }

    /**
     * start to load configuration and lifeCycle
     */
    await this.containerLoader.refresh();
    await this.afterInitialize(options);
  }

  public getApplicationContext(): IMidwayContainer {
    return this.containerLoader.getApplicationContext();
  }

  public getConfiguration(key?: string): any {
    return this.getApplicationContext()
      .getConfigService()
      .getConfiguration(key);
  }

  public getCurrentEnvironment() {
    return this.getApplicationContext()
      .getEnvironmentService()
      .getCurrentEnvironment();
  }

  public abstract getFrameworkType(): MidwayFrameworkType;

  public abstract getApplication(): APP;

  public abstract run(): Promise<void>;

  public async stop(): Promise<void> {
    await this.beforeStop();
    await this.containerLoader.stop();
  }

  protected defineApplicationProperties(app: APP): APP {
    return Object.assign(app, {
      getBaseDir: () => {
        return this.baseDir;
      },

      getAppDir: () => {
        return this.appDir;
      },

      getEnv: () => {
        return this.getApplicationContext()
          .getEnvironmentService()
          .getCurrentEnvironment();
      },

      getConfig: (key?: string) => {
        return this.getApplicationContext()
          .getConfigService()
          .getConfiguration(key);
      },

      getFrameworkType: () => {
        return this.getFrameworkType();
      },

      getProcessType: () => {
        return MidwayProcessTypeEnum.APPLICATION;
      },
    });
  }

  protected async beforeStop(): Promise<void> {}

  protected async beforeInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async beforeDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected async afterDirectoryLoad(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void> {}

  protected abstract async afterInitialize(
    options: Partial<IMidwayBootstrapOptions>
  ): Promise<void>;
}
