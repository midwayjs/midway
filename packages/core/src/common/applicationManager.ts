import {
  IMidwayApplication,
  IMidwayFramework,
  FrameworkType,
  ScopeEnum,
} from '../interface';
import { Provide, Scope } from '../decorator';

@Provide()
@Scope(ScopeEnum.Singleton)
export class MidwayApplicationManager {
  private globalFrameworkMap = new Map<
    string,
    IMidwayFramework<any, any, any>
  >();

  private globalFrameworkTypeMap = new WeakMap<
    FrameworkType,
    IMidwayFramework<any, any, any>
  >();

  public addFramework(namespace, framework: IMidwayFramework<any, any, any>) {
    this.globalFrameworkMap.set(namespace, framework);
    if (framework['getFrameworkType']) {
      this.globalFrameworkTypeMap.set(
        framework['getFrameworkType'](),
        framework
      );
    }
  }

  public getFramework(namespaceOrFrameworkType: string | FrameworkType) {
    if (typeof namespaceOrFrameworkType === 'string') {
      if (this.globalFrameworkMap.has(namespaceOrFrameworkType)) {
        return this.globalFrameworkMap.get(namespaceOrFrameworkType);
      }
    } else {
      if (this.globalFrameworkTypeMap.has(namespaceOrFrameworkType)) {
        return this.globalFrameworkTypeMap.get(namespaceOrFrameworkType);
      }
    }
  }

  public getApplication(
    namespaceOrFrameworkType: string | FrameworkType
  ): IMidwayApplication {
    if (typeof namespaceOrFrameworkType === 'string') {
      if (this.globalFrameworkMap.has(namespaceOrFrameworkType)) {
        return this.globalFrameworkMap
          .get(namespaceOrFrameworkType)
          .getApplication();
      }
    } else {
      if (this.globalFrameworkTypeMap.has(namespaceOrFrameworkType)) {
        return this.globalFrameworkTypeMap
          .get(namespaceOrFrameworkType)
          .getApplication();
      }
    }
  }

  public getApplications(
    namespaces?: Array<string | FrameworkType>
  ): IMidwayApplication[] {
    if (!namespaces) {
      return Array.from(this.globalFrameworkMap.values()).map(framework => {
        return framework.getApplication();
      });
    } else {
      return namespaces
        .map(namespace => {
          return this.getApplication(namespace);
        })
        .filter(app => {
          return !!app;
        });
    }
  }
}
