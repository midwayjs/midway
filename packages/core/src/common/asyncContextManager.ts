/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export interface AsyncContext {
  /**
   * Get a value from the context.
   *
   * @param key key which identifies a context value
   */
  getValue(key: symbol): unknown;

  /**
   * Create a new context which inherits from this context and has
   * the given key set to the given value.
   *
   * @param key context key for which to set the value
   * @param value value to set for the given key
   */
  setValue(key: symbol, value: unknown): AsyncContext;

  /**
   * Return a new context which inherits from this context but does
   * not contain a value for the given key.
   *
   * @param key context key for which to clear a value
   */
  deleteValue(key: symbol): AsyncContext;
}

export interface AsyncContextManager {
  /**
   * Get the current active context
   */
  active(): AsyncContext;

  /**
   * Run the fn callback with object set as the current active context
   * @param context Any object to set as the current active context
   * @param fn A callback to be immediately run within a specific context
   * @param thisArg optional receiver to be used for calling fn
   * @param args optional arguments forwarded to fn
   */
  with<A extends unknown[], F extends (...args: A) => ReturnType<F>>(
    context: AsyncContext,
    fn: F,
    thisArg?: ThisParameterType<F>,
    ...args: A
  ): ReturnType<F>;

  /**
   * Bind an object as the current context (or a specific one)
   * @param [context] Optionally specify the context which you want to assign
   * @param target Any object to which a context need to be set
   */
  bind<T>(context: AsyncContext, target: T): T;

  /**
   * Enable context management
   */
  enable(): this;

  /**
   * Disable context management
   */
  disable(): this;
}

class AsyncBaseContext implements AsyncContext {
  private _currentContext!: Map<symbol, unknown>;

  /**
   * Construct a new context which inherits values from an optional parent context.
   *
   * @param parentContext a context from which to inherit values
   */
  constructor(parentContext?: Map<symbol, unknown>) {
    // for minification
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this;

    self._currentContext = parentContext ? new Map(parentContext) : new Map();

    self.getValue = (key: symbol) => self._currentContext.get(key);

    self.setValue = (key: symbol, value: unknown): AsyncContext => {
      const context = new AsyncBaseContext(self._currentContext);
      context._currentContext.set(key, value);
      return context;
    };

    self.deleteValue = (key: symbol): AsyncContext => {
      const context = new AsyncBaseContext(self._currentContext);
      context._currentContext.delete(key);
      return context;
    };
  }

  /**
   * Get a value from the context.
   *
   * @param key key which identifies a context value
   */
  public getValue!: (key: symbol) => unknown;

  /**
   * Create a new context which inherits from this context and has
   * the given key set to the given value.
   *
   * @param key context key for which to set the value
   * @param value value to set for the given key
   */
  public setValue!: (key: symbol, value: unknown) => AsyncContext;

  /**
   * Return a new context which inherits from this context but does
   * not contain a value for the given key.
   *
   * @param key context key for which to clear a value
   */
  public deleteValue!: (key: symbol) => AsyncContext;
}

/** The root context is used as the default parent context when there is no active context */
export const ASYNC_ROOT_CONTEXT: AsyncContext = new AsyncBaseContext();

export class NoopContextManager implements AsyncContextManager {
  active(): AsyncContext {
    return ASYNC_ROOT_CONTEXT;
  }

  with<A extends unknown[], F extends (...args: A) => ReturnType<F>>(
    _context: AsyncContext,
    fn: F,
    thisArg?: ThisParameterType<F>,
    ...args: A
  ): ReturnType<F> {
    return fn.call(thisArg, ...args);
  }

  bind<T>(_context: AsyncContext, target: T): T {
    return target;
  }

  enable(): this {
    return this;
  }

  disable(): this {
    return this;
  }
}
