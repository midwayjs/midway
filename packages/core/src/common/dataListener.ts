import { Destroy, Init } from '../decorator';

export abstract class DataListener<T> {
  private innerData: T;

  @Init()
  protected async init() {
    this.innerData = await this.initData();
    await this.onData(this.setData.bind(this));
  }

  abstract initData(): T | Promise<T>;
  abstract onData(callback: (data: T) => void);

  protected setData(data: T): void {
    this.innerData = data;
  }

  public getData(): T {
    return this.innerData;
  }

  @Destroy()
  public async stop() {
    await this.destroyListener();
  }

  protected async destroyListener() {}
}
