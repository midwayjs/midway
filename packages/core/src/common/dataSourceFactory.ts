/**
 * 数据源工厂实现
 */
import { extend } from '../util/extend';
// import * as assert from 'assert';
import { MidwayCommonError } from '../error';

export abstract class DataSourceManager<T> {
  protected dataSource: Map<string, T> = new Map();
  protected options = {};
  protected dataSourceNameMap = new Map();

  protected async initDataSource(options: any = {}): Promise<void> {
    this.options = options;
    if (options.dataSource) {
      for (const id of Object.keys(options.dataSource)) {
        await this.createDataSource(options. [id], id);
      }
    } else {
      throw new MidwayCommonError('DataSource must set options.dataSource.');
    }
  }

  public getDataSource(dataSourceName: string) {
    return this.dataSource.get(dataSourceName);
  }

  public getDataSourceByGroup(dataSourceName: string): T[] {
    return this.dataSource.get(dataSourceName);
  }


  public hasDataSource(id: string): boolean {
    return this.dataSource.has(id);
  }

  public async createInstance(config, clientName?): Promise<T | void> {
    // options.default will be merge in to options.clients[id]
    config = extend(true, {}, this.options['default'], config);
    const client = await this.createDataSource(config, clientName);
    if (client) {
      if (clientName) {
        this.dataSource.set(clientName, client);
      }
      return client;
    }
  }

  public abstract getName(): string;
  protected abstract createDataSource(
    config,
    dataSourceName: string
  ): Promise<T | void> | (T | void);
  protected async destroyDataSource(client: T): Promise<void> {}

  public async stop(): Promise<void> {
    for (const value of this.dataSource.values()) {
      await this.destroyDataSource(value);
    }
  }

  public bindEntityWithDataSource(entity, dataSourceOrGroup: string) {
    this.dataSourceNameMap.set();
  }
}
