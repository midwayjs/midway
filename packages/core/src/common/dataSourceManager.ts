/**
 * 数据源管理器实现
 */
import { extend } from '../util/extend';
import { MidwayCommonError, MidwayParameterError } from '../error';
import { run } from '@midwayjs/glob';
import { join } from 'path';
import { Types } from '@midwayjs/decorator';
import { DEFAULT_PATTERN, IGNORE_PATTERN } from '../interface';

export abstract class DataSourceManager<T> {
  protected dataSource: Map<string, T> = new Map();
  protected options = {};
  protected modelMapping = new WeakMap();

  protected async initDataSource(options: any, appDir: string): Promise<void> {
    this.options = options;
    if (!options.dataSource) {
      throw new MidwayParameterError(
        '[DataSourceManager] must set options.dataSource.'
      );
    }

    for (const dataSourceName in options.dataSource) {
      const dataSourceOptions = options.dataSource[dataSourceName];
      if (dataSourceOptions['entities']) {
        const entities = new Set();
        // loop entities and glob files to model
        for (const entity of dataSourceOptions['entities']) {
          if (typeof entity === 'string') {
            // string will be glob file
            const models = globModels(entity, appDir);
            for (const model of models) {
              entities.add(model);
              this.modelMapping.set(model, dataSourceName);
            }
          } else {
            // model will be add to array
            entities.add(entity);
            this.modelMapping.set(entity, dataSourceName);
          }
        }
        dataSourceOptions['entities'] = Array.from(entities);
      }
      // create data source
      const opts: CreateDataSourceInstanceOptions = {
        cacheInstance: options.cacheInstance, // will default true
        validateConnection: options.validateConnection,
      };
      await this.createInstance(dataSourceOptions, dataSourceName, opts);
    }
  }

  /**
   * get a data source instance
   * @param dataSourceName
   */
  public getDataSource(dataSourceName: string) {
    return this.dataSource.get(dataSourceName);
  }

  /**
   * check data source has exists
   * @param dataSourceName
   */
  public hasDataSource(dataSourceName: string): boolean {
    return this.dataSource.has(dataSourceName);
  }

  public getDataSourceNames() {
    return Array.from(this.dataSource.keys());
  }

  /**
   * check the data source is connected
   * @param dataSourceName
   */
  public async isConnected(dataSourceName: string): Promise<boolean> {
    const inst = this.getDataSource(dataSourceName);
    return inst ? this.checkConnected(inst) : false;
  }

  public async createInstance(
    config: any,
    clientName: any,
    options?: CreateDataSourceInstanceOptions
  ): Promise<T | void> {
    const cache =
      options && typeof options.cacheInstance === 'boolean'
        ? options.cacheInstance
        : true;
    const validateConnection = (options && options.validateConnection) || false;

    // options.clients[id] will be merged with options.default
    const configNow = extend(true, {}, this.options['default'], config);
    const client = await this.createDataSource(configNow, clientName);
    if (cache && clientName && client) {
      this.dataSource.set(clientName, client);
    }

    if (validateConnection) {
      if (!client) {
        throw new MidwayCommonError(
          `[DataSourceManager] ${clientName} initialization failed.`
        );
      }

      const connected = await this.checkConnected(client);
      if (!connected) {
        throw new MidwayCommonError(
          `[DataSourceManager] ${clientName} is not connected.`
        );
      }
    }

    return client;
  }

  /**
   * get data source name by model or repository
   * @param modelOrRepository
   */
  public getDataSourceNameByModel(modelOrRepository: any): string | undefined {
    return this.modelMapping.get(modelOrRepository);
  }

  public abstract getName(): string;
  protected abstract createDataSource(
    config,
    dataSourceName: string
  ): Promise<T | void> | (T | void);
  protected abstract checkConnected(dataSource: T): Promise<boolean>;
  protected abstract destroyDataSource(dataSource: T): Promise<void>;

  public async stop(): Promise<void> {
    const arr = Array.from(this.dataSource.values());
    await Promise.all(
      arr.map(dbh => {
        return this.destroyDataSource(dbh);
      })
    );
    this.dataSource.clear();
  }
}

export function globModels(globString: string, appDir: string) {
  let cwd;
  let pattern;

  if (globString.endsWith('**')) {
    // 去掉尾部的 **，因为 glob 会自动添加
    globString = globString.slice(0, -2);
  }

  if (/\*/.test(globString)) {
    cwd = appDir;
    pattern = [...DEFAULT_PATTERN.map(p => join(globString, p))];
    pattern.push(globString);
  } else {
    pattern = [...DEFAULT_PATTERN];
    cwd = join(appDir, globString);
  }

  const models = [];
  // string will be glob file
  const files = run(pattern, {
    cwd,
    ignore: IGNORE_PATTERN,
  });
  for (const file of files) {
    const exports = require(file);
    if (Types.isClass(exports)) {
      models.push(exports);
    } else {
      for (const m in exports) {
        const module = exports[m];
        if (Types.isClass(module)) {
          models.push(module);
        }
      }
    }
  }
  return models;
}

export interface CreateDataSourceInstanceOptions {
  /**
   * @default false
   */
  validateConnection?: boolean;
  /**
   * @default true
   */
  cacheInstance?: boolean | undefined;
}
