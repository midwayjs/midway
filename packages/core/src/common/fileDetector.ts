import {
  IFileDetector,
  IMidwayContainer,
  IObjectDefinition,
} from '../interface';
import { Types } from '../util/types';
import { run } from '@midwayjs/glob';
import { MidwayDuplicateClassNameError } from '../error';
import { DEFAULT_PATTERN, IGNORE_PATTERN } from '../constants';
import { getProviderName } from '../decorator';

export abstract class AbstractFileDetector<T> implements IFileDetector {
  options: T;
  extraDetectorOptions: T;
  constructor(options) {
    this.options = options;
    this.extraDetectorOptions = {} as T;
  }

  abstract run(container: IMidwayContainer);

  setExtraDetectorOptions(detectorOptions: T) {
    this.extraDetectorOptions = detectorOptions;
  }
}

const DEFAULT_GLOB_PATTERN = ['**/**.tsx'].concat(DEFAULT_PATTERN);
const DEFAULT_IGNORE_PATTERN = [
  '**/logs/**',
  '**/run/**',
  '**/public/**',
  '**/app/view/**',
  '**/app/views/**',
  '**/app/extend/**',
  '**/node_modules/**',
  '**/**.test.ts',
  '**/**.test.js',
  '**/__test__/**',
].concat(IGNORE_PATTERN);

export class DirectoryFileDetector extends AbstractFileDetector<{
  loadDir?: string | string[];
  pattern?: string | string[];
  ignore?: string | string[];
  namespace?: string;
  conflictCheck?: boolean;
}> {
  private duplicateModuleCheckSet = new Map();

  run(container) {
    const loadDirs = []
      .concat(this.options.loadDir || [])
      .concat(this.extraDetectorOptions.loadDir || []);

    for (const dir of loadDirs) {
      const fileResults = run(
        DEFAULT_GLOB_PATTERN.concat(this.options.pattern || []).concat(
          this.extraDetectorOptions.pattern || []
        ),
        {
          cwd: dir,
          ignore: DEFAULT_IGNORE_PATTERN.concat(
            this.options.ignore || []
          ).concat(this.extraDetectorOptions.ignore || []),
        }
      );

      // 检查重复模块
      const checkDuplicatedHandler = (module, options?: IObjectDefinition) => {
        if (
          (this.options.conflictCheck ||
            this.extraDetectorOptions.conflictCheck) &&
          Types.isClass(module)
        ) {
          const name = getProviderName(module);
          if (name) {
            if (this.duplicateModuleCheckSet.has(name)) {
              throw new MidwayDuplicateClassNameError(
                name,
                options.srcPath,
                this.duplicateModuleCheckSet.get(name)
              );
            } else {
              this.duplicateModuleCheckSet.set(name, options.srcPath);
            }
          }
        }
      };

      for (const file of fileResults) {
        const exports = require(file);
        // add module to set
        container.bindClass(exports, {
          namespace: this.options.namespace,
          srcPath: file,
          createFrom: 'file',
          bindHook: checkDuplicatedHandler,
        });
      }
    }

    // check end
    this.duplicateModuleCheckSet.clear();
  }
}

export class CustomModuleDetector extends AbstractFileDetector<{
  modules?: any[];
  namespace?: string;
}> {
  run(container) {
    for (const module of this.options.modules) {
      container.bindClass(module, {
        namespace: this.options.namespace,
        createFrom: 'module',
      });
    }
  }
}
