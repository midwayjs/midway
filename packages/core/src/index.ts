export {
  ObjectIdentifier,
  ObjectDefinitionOptions,
  IManagedInstance,
  ScopeEnum,
  saveClassMetadata,
  attachClassMetadata,
  getClassMetadata,
  saveMethodDataToClass,
  attachMethodDataToClass,
  getMethodDataFromClass,
  listMethodDataFromClass,
  saveMethodMetadata,
  attachMethodMetadata,
  getMethodMetadata,
  savePropertyDataToClass,
  attachPropertyDataToClass,
  getPropertyDataFromClass,
  listPropertyDataFromClass,
  savePropertyMetadata,
  attachPropertyMetadata,
  getPropertyMetadata,
  savePreloadModule,
  listPreloadModule,
  saveModule,
  listModule,
  resetModule,
  clearAllModule,
  getParamNames,
  getProviderId,
  getObjectDefinition,
  classNamed,
  generateProvideId
} from '@midwayjs/decorator';
export * from './interface';
export { ContainerLoader } from './loader';
export { MidwayContainer } from './context/midwayContainer';
export { MidwayRequestContainer } from './context/requestContainer';
export { BaseFramework } from './baseFramework';
export * from './context/providerWrapper';
export * from './common/constants';
export {
  safelyGet,
  safeRequire,
  parsePrefix,
} from './common/util';
export * from './features';
export * from './util/webRouterParam';
export { plainToClass, classToPlain } from 'class-transformer';
