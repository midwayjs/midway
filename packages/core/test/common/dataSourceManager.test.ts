import { DataSourceManager } from '../../src';
import { globModels } from '../../src/common/dataSourceManager';
import { join } from 'path';

describe('test/common/dataSourceManager.test.ts', () => {

  class CustomDataSourceFactory extends DataSourceManager<any> {
    getName() {
      return 'test';
    }
    async init(options) {
      return super.initDataSource(options, __dirname);
    }

    protected createDataSource(config, dataSourceName: string): any {
      config.entitiesLength = config.entities.length;
      return config;
    }

    protected async checkConnected(dataSource: any) {
      return false;
    }

    protected async destroyDataSource(dataSource: any): Promise<void> {
      return
    }
  }

  it('should test base data source', async () => {
    class EntityA {}

    class EntityB {}

    const instance = new CustomDataSourceFactory();
    expect(instance.getName()).toEqual('test');

    await instance.init({
      dataSource: {
        default: {
          host: 'localhost',    //数据库地址,默认本机
          port:'3306',
          dialect: 'mysql',
          pool: {   //连接池设置
            max: 5, //最大连接数
            min: 0, //最小连接数
            idle: 10000
          },
          entities: [EntityA, EntityB]
        },
        test: {
          host: 'localhost',    //数据库地址,默认本机
          port:'3306',
          dialect: 'mysql',
          pool: {   //连接池设置
            max: 5, //最大连接数
            min: 0, //最小连接数
            idle: 10000
          },
          dataSourceGroup: 'bb',
          entities: [EntityA]
        }
      },
    })
    expect(instance.getDataSourceNames()).toEqual(['default', 'test']);
    expect(instance.hasDataSource('default')).toBeTruthy();
    expect(instance.getDataSource('default')).toMatchSnapshot();
    expect(instance.getDataSource('default').entitiesLength).toEqual(2);
    expect(instance.getDataSource('test')).toMatchSnapshot();
    expect(instance.getDataSource('test').entitiesLength).toEqual(1);
    expect(instance.getDataSource('fff')).not.toBeDefined();

    expect(await instance.isConnected('default')).toBeFalsy();

    expect(instance.getDataSourceNameByModel(EntityB)).toEqual('default');
    expect(instance.getDataSourceNameByModel(EntityA)).toEqual('test');

    await instance.stop();
    expect(instance.getDataSourceNames()).toEqual([]);
  });

  it('should test glob model', function () {
    let result = globModels('dd', __dirname);
    expect(result).toEqual([]);

    result = globModels('abc', __dirname);
    expect(result.length).toEqual(4);
  });

  it('should test glob model with pattern string', function () {
    let result = globModels('**/bcd/**', join(__dirname, 'glob_dir_pattern'));
    expect(result.length).toEqual(1);
  });

  it('should test with glob model', async () => {
    class EntityA {}

    const instance = new CustomDataSourceFactory();
    expect(instance.getName()).toEqual('test');

    await instance.init({
      dataSource: {
        test: {
          host: 'localhost',    //数据库地址,默认本机
          port:'3306',
          dialect: 'mysql',
          pool: {   //连接池设置
            max: 5, //最大连接数
            min: 0, //最小连接数
            idle: 10000
          },
          entities: [EntityA, EntityA, '/abc']
        },
      },
    })
    expect(instance.getDataSourceNames()).toEqual(['test']);
    expect(instance.getDataSource('test')).toMatchSnapshot();
  });

  it('should createInstance() without cacheInstance (default true)', async () => {
    class EntityA {}

    const instance = new CustomDataSourceFactory();
    expect(instance.getName()).toEqual('test');


    const clientName ='test'
    const config = {
      host: 'localhost',    //数据库地址,默认本机
      port: '3306',
      dialect: 'mysql',
      pool: {   //连接池设置
        max: 5, //最大连接数
        min: 0, //最小连接数
        idle: 10000
      },
      entities: [EntityA, EntityA, '/abc']
    }

    await instance.createInstance(config, clientName)
    expect(instance.getDataSourceNames()).toEqual([clientName]);
  });

  it('should createInstance() with cacheInstance: true', async () => {
    class EntityA {}

    const instance = new CustomDataSourceFactory();
    expect(instance.getName()).toEqual('test');

    const clientName ='test'
    const config = {
      host: 'localhost',    //数据库地址,默认本机
      port: '3306',
      dialect: 'mysql',
      pool: {   //连接池设置
        max: 5, //最大连接数
        min: 0, //最小连接数
        idle: 10000
      },
      entities: [EntityA, EntityA, '/abc']
    }

    await instance.createInstance(config, clientName, { cacheInstance: true })
    expect(instance.getDataSourceNames()).toEqual([clientName]);
  });

  it('should createInstance() with cacheInstance: false', async () => {
    class EntityA {}

    const instance = new CustomDataSourceFactory();
    expect(instance.getName()).toEqual('test');

    const clientName ='test'
    const config = {
      host: 'localhost',    //数据库地址,默认本机
      port: '3306',
      dialect: 'mysql',
      pool: {   //连接池设置
        max: 5, //最大连接数
        min: 0, //最小连接数
        idle: 10000
      },
      entities: [EntityA, EntityA, '/abc']
    }

    await instance.createInstance(config, clientName, { cacheInstance: false })
    expect(instance.getDataSourceNames()).not.toEqual([clientName]);
    expect(instance.getDataSource(clientName)).toBeUndefined();
  });

  it('should test will got error when no data source', async () => {
    const instance = new CustomDataSourceFactory();
    let e;
    try {
      await instance.init({});
    } catch (err) {
      e = err;
    }

    expect(e).toBeDefined();
  });
});
