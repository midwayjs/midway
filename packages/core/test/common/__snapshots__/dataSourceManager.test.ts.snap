// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/common/dataSourceManager.test.ts should test base data source 1`] = `
{
  "dialect": "mysql",
  "entities": [
    [Function],
    [Function],
  ],
  "entitiesLength": 2,
  "host": "localhost",
  "pool": {
    "idle": 10000,
    "max": 5,
    "min": 0,
  },
  "port": "3306",
}
`;

exports[`test/common/dataSourceManager.test.ts should test base data source 2`] = `
{
  "dataSourceGroup": "bb",
  "dialect": "mysql",
  "entities": [
    [Function],
  ],
  "entitiesLength": 1,
  "host": "localhost",
  "pool": {
    "idle": 10000,
    "max": 5,
    "min": 0,
  },
  "port": "3306",
}
`;

exports[`test/common/dataSourceManager.test.ts should test with glob model 1`] = `
{
  "dialect": "mysql",
  "entities": [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "entitiesLength": 5,
  "host": "localhost",
  "pool": {
    "idle": 10000,
    "max": 5,
    "min": 0,
  },
  "port": "3306",
}
`;
